### PR TITLE
#171380752 one way trip request 

### PR DESCRIPTION
**Pull request by since9629**
#### What does this PR do?
Creates API endpoint to receive _one-way trip requests_ from users
#### Description of Task to be completed?
- Create a request model and its migrations
- create API endpoint at **POST /api/trips**
- Create `Authorization middleware` to extract and check requester info from `Authorization token` obtained after login
#### How should this be manually tested?
- After cloning the project repo, run `npm i` inside the cloned repo folder
- Make sure to rename .env.example to .env, 
- In .env file, provide all values for variables under `#Development database`  and `#General`
- Finally, run the following commands respectively:
`git fetch`
`git checkout ft-request-one-way-trip-171380752`
`npm i`
`npx sequelize-cli db:create`
`npx sequelize-cli db:migrate`
`npm run start`
- If the above steps have been followed correctly, you should see a new users and trips table created in your database
- To create a one way trip request, send the following data in your request body as JSON object to `POST http://<your_server_hostname>:<server_port>/api/trips` endpoint:
```json
    travelFrom: 'kigali',
    travelTo: 'Butare',
    travelDate: '2020-03-14T15:12:22.360Z',
    travelReason: 'business meeting',
    travelType: 'one-way',
    accommodation: true,
```
#### Any background context you want to provide?
You need to create an account and use the obtained token as an `authorization header` while creating _one way trip request_ 
#### What are the relevant pivotal tracker stories?
[#171380752](https://www.pivotaltracker.com/n/projects/2435482/stories/171380752)
#### Screenshots (if appropriate)
Example request:
req

Example response:
res

### CODE REVIEW

**Review by descholar-ceo — COMMENTED** [review]
Great work

**Review by muchristian — COMMENTED** [review]
Patrick you did great job, but i didn't see where you've added relation between user and trip. take a look on associations/relations in Sequelize

**Review by Muhire-Josue — COMMENTED** [review]
Great work @karamuka, it's very impressive. However, I do have a few comments that I think will make your work even better
- could you make the `storyID` on the PR description a link to that story in PT?
- could you add validation to prevent Duplication of one-way trip? I have seen that if you send the same data over and over again it actually duplicates the same trip in the database. I think it'd be better if the user can't create the same trip more than once.
- could you include screenshots of request and response on postman in the PR description? it gives a loose idea about what to expect while testing this endpoint in postman.
- can you add the optional 3000 on the port number in the server? this is the error am getting and this is how I fixed it, perhaps you should consider to make this changes as well so that when someone else fetches your code doesn't have to add it themselves.
server
port

**Comment by since9629** [comment]
> 
> 
> Great work @karamuka, it's very impressive. However, I do have a few comments that I think will make your work even better
> 
>     * could you make the `storyID` on the PR description a link to that story in PT?
> 
>     * could you add validation to prevent Duplication of one-way trip? I have seen that if you send the same data over and over again it actually duplicates the same trip in the database. I think it'd be better if the user can't create the same trip more than once.
> 
>     * could you include screenshots of request and response on postman in the PR description? it gives a loose idea about what to expect while testing this endpoint in postman.
> 
>     * can you add the optional 3000 on the port number in the server? this is the error am getting and this is how I fixed it, perhaps you should consider to make this changes as well so that when someone else fetches your code doesn't have to add it themselves.
>       server
>       port

Thanks @Muhire-Josue for your feedback, however if I provide a port directly into code, it will decrease code coverage. make sure you specify PORT in .env file as I suggested in PR description

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
@karamuka Special congratulations to you, this is great, I personally liked the middleware you created for checking if a user has logged in, that is great and I should recommend everyone to use that for protected resources, though in that middleware you could also use the `findUserByEmailOrUsername()` function found in authentication.service.js module which is doing the same thing as the `findUserByEmail()` the one you used. 

I have a few suggestions for you, that should be mandatory to take as considerations:

### A.Considerations:
1. I saw that a user can't create two or more one-way trips on the same day, is that applicable in real life? I don't think so, imagine a scenario where an employee can be required to travel from two close cities like Kigali-Muhanga (his office is in Kigali), and the aim of the trip is supposed to end before noon, so afternoon he/she comes back to the office and he/she goes to another trip to Nyamata, don't you think after the first trip, he/she will need to request the second trip?
=> In that case, one user will need to request more than one trip.

2. A user who is not yet verified can create a trip.

### B. Suggestions:
1. You can create a validation for a trip, to check if it exists or not, this will be easy for you if you think about specific id for a specific trip.

2. You can also check if a user is verified or not before allowing him/her to request a trip

**Review by Muhire-Josue — APPROVED** [review]
LGTM

**Comment by since9629** [comment]
> 
> 
> @karamuka Special congratulations to you, this is great, I personally liked the middleware you created for checking if a user has logged in, that is great and I should recommend everyone to use that for protected resources, though in that middleware you could also use the `findUserByEmailOrUsername()` function found in authentication.service.js module which is doing the same thing as the `findUserByEmail()` the one you used.
> 
> I have a few suggestions for you, that should be mandatory to take as considerations:
> ### A.Considerations:
> 
>     1. I saw that a user can't create two or more one-way trips on the same day, is that applicable in real life? I don't think so, imagine a scenario where an employee can be required to travel from two close cities like Kigali-Muhanga (his office is in Kigali), and the aim of the trip is supposed to end before noon, so afternoon he/she comes back to the office and he/she goes to another trip to Nyamata, don't you think after the first trip, he/she will need to request the second trip?
>        => In that case, one user will need to request more than one trip.
> 
>     2. A user who is not yet verified can create a trip.
> 
> 
> ### B. Suggestions:
> 
>     1. You can create a validation for a trip, to check if it exists or not, this will be easy for you if you think about specific id for a specific trip.
> 
>     2. You can also check if a user is verified or not before allowing him/her to request a trip

Thank you for your feedback @descholar-ceo, I have implemented all of it except for _creating trip requests multiple times for the same timestamp_. You see if a requester requests a trip on 25-05-2020 at 10h00, and repeats it again it'd be a bug because of duplication.

**Review by muchristian — APPROVED** [review]

**Review by the22mastermind — COMMENTED** [review]
Great job @karamuka However check out the comments I left you. Keep it up.

File: src/controllers/trip.controller.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  /**
+   * @param {Request} req Node/express request
+   * @param {Response} res Node/express response
+   * @param {NextFunction} next Node/Express Next callback function
+   * @returns {Object} Custom response with created trip details
+   * @description One way trip request handler
+   */
+  static async createOneWayTripRequest(req, res, next) {
+    try {
+      const schema = Joi.object({
```
Comment (the22mastermind) [inline src/controllers/trip.controller.js]: Can you handle these validations in middlewares?

File: src/index.js [hunk 2]
```diff
@@ -7,12 +7,16 @@ import { serve, setup } from 'swagger-ui-express';
 import '@babel/polyfill';
 import swaggerSpecs from '../public/api-docs/swagger.json';
 import allRoutes from './routes/index';
+import customMessages from './utils/customMessages';
+import statusCodes from './utils/statusCodes';
 
 dotenv.config();
 
 const app = express();
-const port = process.env.PORT || 3000;
```
Comment (the22mastermind) [inline src/index.js:15]: Why did you remove the `|| 3000`? Could you put it back please?

File: src/middlewares/authentication.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+   * @description Checks existence of a token in the Authorization request header,
+   * decodes it, and finaly adds the decoded data
+   * to Node/Express's Request Object as sessionUser property
+   */
+  static async isCurrentUserLoggedIn(req, res, next) {
+    let token = req.get('authorization');
+    if (!token) {
+        return errorResponse(res, unAuthorized, tokenAbsent);
+    }
+    token = token.split(' ').pop();
```
Comment (the22mastermind) [inline src/middlewares/authentication.js:34]: Can you make sure you are using the OAuth 2.0 standard format of authorization header?

File: src/utils/responseHandlers.js [hunk 4]
```diff
@@ -3,12 +3,14 @@
  * @param {integer} code
  * @param {string} message
  * @param {string} token
+ * @param {string} data
  * @returns {object} response
  * @description Returns a successful response
  */
-const successResponse = async (res, code, message, token) => res.status(code).json({
+const successResponse = (res, code, message, token, data = null) => res.status(code).json({
```
Comment (the22mastermind) [inline src/utils/responseHandlers.js:10]: Can you use the `...rest` operator when passing additional parameters to res, code, and message?

File: src/index.js (hunk 2 above)
Comment (descholar-ceo) [reply src/index.js:15]: @the22mastermind this port number is supposed to be in `.env` file, when you add that value in `env` and you keep this 3000 here, it decreases the coverage. So that is why @karamuka  removed it I think.

File: tests/server.test.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+const { endpointNotFound } = customMessages;
+
+chai.use(chaiHttp);
+chai.should();
+
+describe('Server availability', () => {
+  it('Should return not found', (done) => {
+    chai
+      .request(server)
+      .get('/non-existent-endpoint')
```
Comment (the22mastermind) [inline tests/server.test.js]: Is this endpoint part of a user making a one-way trip request?

**Comment by descholar-ceo** [comment]
> > @karamuka Special congratulations to you, this is great, I personally liked the middleware you created for checking if a user has logged in, that is great and I should recommend everyone to use that for protected resources, though in that middleware you could also use the `findUserByEmailOrUsername()` function found in authentication.service.js module which is doing the same thing as the `findUserByEmail()` the one you used.
> > I have a few suggestions for you, that should be mandatory to take as considerations:
> > ### A.Considerations:
> > ```
> > 1. I saw that a user can't create two or more one-way trips on the same day, is that applicable in real life? I don't think so, imagine a scenario where an employee can be required to travel from two close cities like Kigali-Muhanga (his office is in Kigali), and the aim of the trip is supposed to end before noon, so afternoon he/she comes back to the office and he/she goes to another trip to Nyamata, don't you think after the first trip, he/she will need to request the second trip?
> >    => In that case, one user will need to request more than one trip.
> > 
> > 2. A user who is not yet verified can create a trip.
> > ```
> > 
> > 
> > ### B. Suggestions:
> > ```
> > 1. You can create a validation for a trip, to check if it exists or not, this will be easy for you if you think about specific id for a specific trip.
> > 
> > 2. You can also check if a user is verified or not before allowing him/her to request a trip
> > ```
> 
> Thank you for your feedback @descholar-ceo, I have implemented all of it except for _creating trip requests multiple times for the same timestamp_. You see if a requester requests a trip on 25-05-2020 at 10h00, and repeats it again it'd be a bug because of duplication.

@karamuka Special congrats, your work is excellent. You implemented all the feedback you got, and now a user can request a trip if he's only registered, and in different timestamp he/she can request more than one trip. I like it, keep it up, my brother!

**Review by descholar-ceo — APPROVED** [review]

File: src/index.js (hunk 2 above)
Comment (the22mastermind) [reply src/index.js:15]: That is strange because on the previous PRs it didn't decrease the coverage.

File: src/middlewares/authentication.js (hunk 3 above)
Comment (since9629) [reply src/middlewares/authentication.js:34]: could you elaborate @the22mastermind

File: src/utils/responseHandlers.js (hunk 4 above)
Comment (since9629) [reply src/utils/responseHandlers.js:10]: that'd be good, but for this function I had to specify exactly which param I need to hold res data

**Comment by kipropbrian** [comment]
Instead of Timestamp '2020-03-14T15:12:22.360Z', can it be just date '2020-03-14', are the locations `Kigali` in the DB or are they strings from the user? Is travel-direction enum or string

File: src/utils/customMessages.js [hunk 6]
```diff
@@ -20,7 +20,16 @@ const messages = {
...
   invalidAddress: 'Address should consists of alphabets and at least 3 minimum characters, without whitespace or special characters',
-  verifyMessage: 'You have now verified your account successfully'
+  verifyMessage: 'You have now verified your account successfully',
+  loginPasswordAndEmailEmpty: 'Please enter your email and your password',
+  oneWayTripRequestFailed: 'trip request request failed',
+  oneWayTripRequestCreated: 'trip request successfuly submitted and waiting for approval',
+  serverError: 'Sorry, something went wrong on our end, please try again later',
+  tokenAbsent: 'access denied, you must login to use this resource',
+  tokenVerifyFailed: 'sorry, we were unable to verify your identity',
+  endpointNotFound: 'requested resource was not found',
```
Comment (mifeille) [inline src/utils/customMessages.js]: Please make this error message more specific.

File: src/utils/customMessages.js [hunk 7]
```diff
@@ -20,7 +20,16 @@ const messages = {
...
+  invalidTripRequestData: 'submitted information is invalid, or you have submitted the same trip request twice',
```
Comment (mifeille) [inline src/utils/customMessages.js]: Error messages have to be specific, users need to know what they are doing wrong. for example, I submitted the form without the travel date and this is the error I am getting if I was a user I would be confused.
You can check how the signup validations for the body are implemented.

File: src/middlewares/validations.js [hunk 8]
```diff
@@ -34,4 +40,30 @@ const validateSignup = user => {
...
+*/
+const validateOneWayTripRequest = async (req, res, next) => {
+  try {
+    const schema = Joi.object({
+      travelFrom: Joi.string().required(),
+      travelTo: Joi.string().required(),
+      travelDate: Joi.date().required(),
+      travelReason: Joi.string().required(),
+      accommodate: Joi.bool().required(),
+      travelDirection: Joi.equal('one-way').required(),
```
Comment (mifeille) [inline src/middlewares/validations.js]: when I add spaces at the beginning or at the ending it is refusing the input, same case when I write One-way, and it is a bug.
All these required fields should have specific error messages they throw when not provided.

File: src/database/migrations/20200312122638-create-trip.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    },
+    travelReason: {
+      allowNull: false,
+      type: Sequelize.STRING
+    },
+    travelType: {
+      allowNull: false,
+      type: Sequelize.STRING
+    },
+    isApproved: {
```
Comment (mifeille) [inline src/database/migrations/20200312122638-create-trip.js]: The request status should not be a boolean since it can be pending, approved or rejected. if we use a boolean we will never know if it is still pending or if it has been rejected.

File: src/index.js (hunk 2 above)
Comment (since9629) [reply src/index.js:15]: I agree with @de-scholar

**Comment by since9629** [comment]
> 
> 
> Instead of Timestamp '2020-03-14T15:12:22.360Z', can it be just date '2020-03-14', are the locations `Kigali` in the DB or are they strings from the user? Is travel-direction enum or string

- For **timestamp**, I though it'd be better to use both date and time to make it possible for a user to create multiple trip requests in one day, so that's why I used timestamp instead just date, because it contains both date and time
- For **travel-direction**, I have changed it to *travelType* and it is of type string.
- As for **locations**, they are strings from a user

**Comment by since9629** [comment]
> 
> 
> > Instead of Timestamp '2020-03-14T15:12:22.360Z', can it be just date '2020-03-14', are the locations `Kigali` in the DB or are they strings from the user? Is travel-direction enum or string
> 
>     * For timestamp, we though it'd be a good idea to use bot date and time to make it possible for a user to create multiple trip requests in one day, so that's why I used timestamp instead just date
